### ISSUE TITLE
Potential panic in Endpoint::handle_event

### ISSUE THREAD

**Issue by Matthias247** [quinn-rs/quinn#1018]
Just observed this panic in the [server_stateless_reset](https://github.com/quinn-rs/quinn/blob/3acf0f29b696f2cec00ff19e8c730026ed2938e8/quinn-proto/src/tests/mod.rs#L129) unit-test:

~~~
Feb 13 22:00:25.184 TRACE server: quinn_proto::connection: connection closed
thread 'tests::server_stateless_reset' panicked at 'index out of bounds: the len is 0 but the index is 0', C:\Users\matth\.cargo\registry\src\github.com-1ecc6299db9ec823\slab-0.4.2\src\lib.rs:668:38
stack backtrace:
  7: quinn_proto::endpoint::Endpoint<quinn_proto::crypto::rustls::TlsSession>::handle_event<quinn_proto::crypto::rustls::TlsSession>
             at .\src\endpoint.rs:136
   8: quinn_proto::tests::util::TestEndpoint::drive
             at .\src\tests\util.rs:278
~~~

This points to [Endpoint::handle_event](https://github.com/quinn-rs/quinn/blob/3acf0f29b696f2cec00ff19e8c730026ed2938e8/quinn-proto/src/endpoint.rs#L134), where the `Drained` handler tries to remove something from a `Slab`, but `Slab` will panic if the key is not known. 

The test failing is likely caused by some other changes I made. However I think those are unrelated to the panic, and might just have uncovered this state. Not sure if this is simply missing a `if self.connections.contains(ch.0)` check, or whether there is an invariant that should lead this state to never be reached.

**Comment by Matthias247** [comment]
FWIW, this happened after a lot of PTO timers fired and finally the idle timer was reached:

````
Feb 13 22:08:47.864 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=12 space=Data
Feb 13 22:08:47.864 TRACE quinn_proto::tests::util: advancing to Instant { t: 40812.4688232s } for client
Feb 13 22:08:47.864 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:08:47.865 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=13 space=Data
Feb 13 22:08:47.877 TRACE quinn_proto::tests::util: advancing to Instant { t: 40814.2768232s } for server
Feb 13 22:08:47.878 TRACE server: quinn_proto::connection: timeout timer=Idle
Feb 13 22:08:47.878 TRACE server: quinn_proto::connection: connection closed
```

**Comment by Ralith** [comment]
Hmm, that suggests that `Drained` was received multiple times for the same connection, which should be impossible. I don't think anything else causes removal of connection metadata from the slab. Does it repro consistently?

**Comment by Ralith** [comment]
Maybe a bug in the test infra passing an event somewhere incorrectly, or using the wrong handle?

**Comment by Matthias247** [comment]
Happens consistently on the following branch: https://github.com/quinn-rs/quinn/compare/main...Matthias247:stateless_panic?expand=1

It's mostly a combination of the GSO branch, plus some additional RangeSet experimentation. Doesn't touch the endpoint logic

**Comment by Matthias247** [comment]
Adding a `trace!`: Seems like there are 2 `Drained` events:

```
Feb 13 22:23:35.417 DEBUG client: quinn_proto::connection: got stateless reset
Feb 13 22:23:35.417 TRACE client: quinn_proto::endpoint: Got Drained event for connection 0
Feb 13 22:23:35.418 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.5417172s } for client
Feb 13 22:23:35.418 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.424 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=0 space=Data
Feb 13 22:23:35.437 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.5427172s } for client
Feb 13 22:23:35.437 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.437 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=1 space=Data
Feb 13 22:23:35.454 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.5447172s } for client
Feb 13 22:23:35.467 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.467 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=2 space=Data
Feb 13 22:23:35.468 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.5487172s } for client
Feb 13 22:23:35.468 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.474 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=3 space=Data
Feb 13 22:23:35.485 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.5567172s } for client
Feb 13 22:23:35.486 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.486 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=4 space=Data
Feb 13 22:23:35.486 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.5727172s } for client
Feb 13 22:23:35.493 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.504 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=5 space=Data
Feb 13 22:23:35.518 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.6047172s } for client
Feb 13 22:23:35.519 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.519 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=6 space=Data
Feb 13 22:23:35.519 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.6687172s } for client
Feb 13 22:23:35.526 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.537 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=7 space=Data
Feb 13 22:23:35.537 TRACE quinn_proto::tests::util: advancing to Instant { t: 41691.7967172s } for client
Feb 13 22:23:35.537 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.538 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=8 space=Data
Feb 13 22:23:35.569 TRACE quinn_proto::tests::util: advancing to Instant { t: 41692.0527172s } for client
Feb 13 22:23:35.570 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.570 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=9 space=Data
Feb 13 22:23:35.570 TRACE quinn_proto::tests::util: advancing to Instant { t: 41692.5647172s } for client
Feb 13 22:23:35.576 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.587 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=10 space=Data
Feb 13 22:23:35.600 TRACE quinn_proto::tests::util: advancing to Instant { t: 41693.5887172s } for client
Feb 13 22:23:35.600 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.601 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=11 space=Data
Feb 13 22:23:35.601 TRACE quinn_proto::tests::util: advancing to Instant { t: 41695.6367172s } for client
Feb 13 22:23:35.608 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.619 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=12 space=Data
Feb 13 22:23:35.619 TRACE quinn_proto::tests::util: advancing to Instant { t: 41699.7327172s } for client
Feb 13 22:23:35.619 TRACE client: quinn_proto::connection: timeout timer=LossDetection
Feb 13 22:23:35.620 TRACE client: quinn_proto::connection: PTO fired in_flight=62 count=13 space=Data
Feb 13 22:23:35.627 TRACE quinn_proto::tests::util: advancing to Instant { t: 41701.5407172s } for server
Feb 13 22:23:35.650 TRACE server: quinn_proto::connection: timeout timer=Idle
Feb 13 22:23:35.650 TRACE server: quinn_proto::connection: connection closed
Feb 13 22:23:35.651 TRACE server: quinn_proto::endpoint: Got Drained event for connection 0
thread 'tests::server_stateless_reset' panicked at 'index out of bounds: the len is 0 but the index is 0', C:\Users\matth\.cargo\registry\src\github.com-1ecc6299db9ec823\slab-0.4.2\src\lib.rs:668:38
```

**Comment by Matthias247** [comment]
Seems like [`handle_packet`](https://github.com/quinn-rs/quinn/blob/3acf0f29b696f2cec00ff19e8c730026ed2938e8/quinn-proto/src/connection/mod.rs#L2066-L2071) creates the first `Drained` event. This should stop the close timer.

However that doesn't seem to happen, since it tuns again and [creates another `Drained` event](https://github.com/quinn-rs/quinn/blob/3acf0f29b696f2cec00ff19e8c730026ed2938e8/quinn-proto/src/connection/mod.rs#L804-L807), which is the duplicate.

**Comment by Matthias247** [comment]
oh, the `Idle` timeout expired, not the `close` one. Seems like the connection still lives on after this reset and handles PTO timers. Then after a while it hits the idle timer and gets a double drain.

**Comment by Matthias247** [comment]
Hmm ok, this might be wrong. One `Drained` event is for the server, the other for the client.

**Comment by Matthias247** [comment]
Oh, is this because the test swaps `Endpoint`, but leaves the remains of `TestEndpoint` there - which means the endpoint inherits events that were deemed for the previous instance?

**Comment by Matthias247** [comment]
And I figured out the condition that lead to the error: It's the GSO changes sending 8 datagrams in a `Transmit`, and the test endpoint not handling it. 

However I'm wondering if it should fail that way or whether this is an indicator of something else being wrong. It should maybe retry sending something if the PTO timer really fires?

**Comment by Ralith** [comment]
It looks like events shouldn't persist outside of `TestEndpoint::drive` calls, so if that's happening it indicates a bug in the test infra.

> It should maybe retry sending something if the PTO timer really fires?

Failure to transmit after a PTO fires is a bug in the proto layer. `Connection::ensure_probe_queued` should be getting called and arranging for at least a ping if nothing else is buffered.

**Comment by Ralith** [comment]
I believe this was resolved by #1033. Strange behavior likely resulted from inconsistent state following a non-close packet being transmitted after close. If we do see similar behavior let's open a new, more specific issue.
